### PR TITLE
feat: add onchain delegation

### PR DESCRIPTION
**Pull request by zsluedem**
## Summary by CodeRabbit

- **New Features**
	- Introduced contract delegation functionality for interacting with the TaiyiCore smart contract on Ethereum.
	- Added new configuration options for execution clients and on-chain delegation in the RelayConfig.
	- Enhanced PreconferElection struct with a proposer public key field.
	- Added a mock implementation for delegation to facilitate testing.
	- Integrated new delegation capabilities into the ProposerApi and related components.
	- Updated the API service to utilize the new delegation contract.

- **Bug Fixes**
	- Commented out the inactive `MAX_GATEWAY_ELECTION_SIZE` constant and the `/elect_preconfer` API endpoint.

- **Documentation**
	- Updated tests to reflect new delegation capabilities and ensure compatibility with existing functionalities.

### CODE REVIEW

**Comment by Akagi201** [comment]
We need to create a new branch offchain-delegation to backup the offchain version

File: crates/api/Cargo.toml [hunk 1]
```diff
@@ -40,6 +40,10 @@ redis.workspace = true
 # Ethereum Types
 reth-primitives.workspace = true
 ethereum-consensus.workspace = true
+alloy-sol-types = "0.8.8"
```
Comment (Akagi201) [inline crates/api/Cargo.toml]: move version to workspace Cargo.toml

File: crates/api/src/delegation/contract_delegation.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    BoxTransport,
+    Ethereum,
+>;
+sol! {
+    #[derive(Debug)]
+    struct PreconferElectionRes {
+        bytes validatorPubkey;
+        bytes preconferPubkey;
+        uint256 chainId;
+        address preconferAddress;
```
Comment (Akagi201) [inline crates/api/src/delegation/contract_delegation.rs:33]: What is preconferAddress used for? maybe not needed?

**Comment by Akagi201** [comment]
We may use a indexer to index delegation event on the contract to reduce the need to query contract on every slot.

File: crates/api/src/proposer/api.rs [hunk 3]
```diff
@@ -1313,6 +1274,29 @@ where
 
         // Update duties if applicable
         if let Some(new_duties) = slot_update.new_duties {
+            for duty in new_duties.iter() {
+                match self.delegation_contract.get_preconfer_election(&duty.entry.registration.message.public_key, duty.slot).await {
+                    Ok(Some(elected_preconfer)) => match self.auctioneer.save_new_gateway_election(&elected_preconfer, duty.slot).await {
```
Comment (Akagi201) [inline crates/api/src/proposer/api.rs:1279]: We need to check if the elected_preconfer is changed before saving to db.

File: crates/api/src/delegation/contract_delegation.rs (hunk 2 above)
Comment (zsluedem) [reply crates/api/src/delegation/contract_delegation.rs:33]: Just want to keep it aligned with https://github.com/lu-bann/taiyi/blob/b3dc7487e3c6b7e75b5223bcd36155f93cf74ea3/contracts/src/interfaces/IDelegationContract.sol#L36-L41. I should put comments here .

File: crates/api/src/proposer/api.rs (hunk 3 above)
Comment (zsluedem) [reply crates/api/src/proposer/api.rs:1279]: address this in another pr https://github.com/lu-bann/helix/issues/25

**Comment by zsluedem** [comment]
> We need to create a new branch offchain-delegation to backup the offchain version

https://github.com/lu-bann/helix/tree/offchain-delegation

**Comment by zsluedem** [comment]
close it since onchain delegation is not our goal now

**Pull request closed without merging**